File: packages/types/src/notification.ts
// ─── Notification Types ───────────────────────────────────────────────────────

export type NotificationType =
  | "order_created"
  | "order_confirmed"
  | "order_preparing"
  | "order_ready"
  | "order_picked_up"
  | "order_delivered"
  | "order_cancelled"
  | "payment_succeeded"
  | "payment_failed"
  | "driver_assigned"
  | "promo_offer"
  | "system";

export type NotificationChannel = "push" | "sms" | "email" | "in_app";

export interface Notification {
  id: string;
  user_id: string;
  type: NotificationType;
  channel: NotificationChannel;
  title: string;
  body: string;
  data: Record<string, string> | null; // deep link data
  is_read: boolean;
  sent_at: string | null;
  read_at: string | null;
  created_at: string;
}

export interface DeviceToken {
  id: string;
  user_id: string;
  token: string;
  platform: "ios" | "android" | "web";
  created_at: string;
  updated_at: string;
}

export interface DeviceTokenInput {
  token: string;
  platform: "ios" | "android" | "web";
}

export interface PushNotificationPayload {
  title: string;
  body: string;
  image_url?: string;
  data?: Record<string, string>;
  badge?: number;
  sound?: string;
}

File: services/notification-svc/app/schemas/notification.py
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class DeviceTokenRequest(BaseModel):
    token: str = Field(min_length=10)
    platform: Literal["ios", "android", "web"]


class NotificationResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    user_id: uuid.UUID
    type: str
    channel: str
    title: str
    body: str
    data: dict[str, Any] | None
    is_read: bool
    sent_at: datetime | None
    read_at: datetime | None
    created_at: datetime


class NotificationListResponse(BaseModel):
    items: list[NotificationResponse]
    total: int
    page: int
    limit: int
    total_pages: int
    unread_count: int
